keps/sig-cli/20190805-pod-troubleshooting.md

```diff
@@ -0,0 +1,380 @@
+
+This is achieved by running a new "debug" container in the pod namespaces. Her
+troubleshooting session might resemble:
+
+```
+% kubectl debug -it -m debian neato-5thn0 -- bash
+root@debug-image:~# ps x
+  PID TTY      STAT   TIME COMMAND
+    1 ?        Ss     0:00 /pause
+   13 ?        Ss     0:00 bash
+   26 ?        Ss+    0:00 /neato
+  107 ?        R+     0:00 ps x
+root@debug-image:~# cat /proc/26/root/etc/resolv.conf
+search default.svc.cluster.local svc.cluster.local cluster.local
+nameserver 10.155.240.10
+options ndots:5
+root@debug-image:~# dig @10.155.240.10 neato.svc.cluster.local.
+
+; <<>> DiG 9.9.5-9+deb8u6-Debian <<>> @10.155.240.10 neato.svc.cluster.local.
+; (1 server found)
+;; global options: +cmd
+;; connection timed out; no servers could be reached
+```
+
+Alice discovers that the cluster's DNS service isn't responding.
+
+#### Debugging
+
+Bob is debugging a tricky issue that's difficult to reproduce. He can't
+reproduce the issue with the debug build, so he attaches a debug container to
+one of the pods exhibiting the problem:
+
+```
+% kubectl debug -it --image=gcr.io/neato/debugger neato-5x9k3 -- sh
+Defaulting container name to debug.
+/ # ps x
+PID   USER     TIME   COMMAND
+    1 root       0:00 /pause
+   13 root       0:00 /neato
+   26 root       0:00 sh
+   32 root       0:00 ps x
+/ # gdb -p 13
+...
+```
+
+He discovers that he needs access to the actual container, which he can achieve
+by installing busybox into the target container:
+
+```
+root@debug-image:~# cp /bin/busybox /proc/13/root
+root@debug-image:~# nsenter -t 13 -m -u -p -n -r /busybox sh
+
+
+BusyBox v1.22.1 (Debian 1:1.22.0-9+deb8u1) built-in shell (ash)
+Enter 'help' for a list of built-in commands.
+
+/ # ls -l /neato
+-rwxr-xr-x    2 0        0           746888 May  4  2016 /neato
+```
+
+Note that running the commands referenced above requires `CAP_SYS_ADMIN` and
+`CAP_SYS_PTRACE`.
+
+This scenario also requires process namespace sharing which is not available
+on Windows.
+
+#### Automation
+
+Carol is a security engineer tasked with running security audits across all of
+her company's running containers. Even though her company has no standard base
+image, she's able to audit all containers using:
+
+```
+% for pod in $(kubectl get -o name pod); do
+    kubectl debug -m gcr.io/neato/security-audit -p $pod /security-audit.sh
+  done
+```
+
+#### Technical Support
+
+Dan's team provides support for his company's multi-tenant cluster. He can
+access the Kubernetes API (as a viewer) on behalf of the users he's supporting,
+but he does not have administrative access to nodes or a say in how the
+application image is constructed. When someone asks for help, Dan's first step
+is to run his team's autodiagnose script:
+
+```
+% kubectl debug --image=k8s.gcr.io/autodiagnose nginx-pod-1234
+```
+
+### Implementation Details/Notes/Constraints
+
+1.  There are no guaranteed resources for ad-hoc troubleshooting. If
```

Comment:
That is rather undesirable effect, since it may cause an application downtime. I think we'll need to be explicit about that in the command's help.

Reply:
I'm even inclined to move this to risks section.

Reply:
Moved to risks and added that we'll be able to mitigate this once support for resizing pods has been added.